gobblin-config-management/src/main/java/gobblin/config/configstore/impl/BaseHdfsConfigStore.java

```diff
@@ -0,0 +1,214 @@
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.log4j.Logger;
+
+import com.google.common.io.Closer;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+
+/**
+ * The BaseHdfsConfigStore implements basic functions from ConfigStore which is related to HDFS
+ * @author mitu
+ *
+ */
+public abstract class BaseHdfsConfigStore implements ConfigStore {
+
+  private static final Logger LOG = Logger.getLogger(BaseHdfsConfigStore.class);
+  public static final String CONFIG_FILE_NAME = "main.conf";
+
+  private final URI storeURI;
+  private final Path location;
+  private final String currentVersion;
+  protected final FileSystem fs;
+  protected final VersionComparator<String> vc;
+
+  public BaseHdfsConfigStore(URI root) {
+    this(root, new SimpleVersionComparator());
+  }
+
+  public BaseHdfsConfigStore(URI root, VersionComparator<String> vc) {
+    try {
+      this.storeURI = root;
+      this.location = new Path(root.getPath());
+      this.fs = FileSystem.get(root, new Configuration());
+    } catch (IOException ioe) {
+      LOG.error("can not initial the file system " + ioe.getMessage(), ioe);
+      throw new RuntimeException(ioe);
+    }
+
+    this.vc = vc;
+    this.currentVersion = this.findCurrentVersion();
+  }
+
+  private String findCurrentVersion() {
+    try {
+      if (this.fs.isFile(this.location)) {
+        throw new RuntimeException(String.format("location %s should be a directory ", this.location));
+      }
+
+      FileStatus[] fileStatus = this.fs.listStatus(this.location);
+      if (fileStatus == null || fileStatus.length == 0) {
+        throw new RuntimeException(String.format("location %s does not have any versions ", this.location));
+      }
+
+      List<String> versions = new ArrayList<String>();
+      for (FileStatus f : fileStatus) {
+        // versions should be directory
+        if (!f.isDir()) {
+          continue;
+        }
+        versions.add(f.getPath().getName());
+      }
+      String res = this.vc.getCurrentVersion(versions);
+
+      if (res == null) {
+        throw new RuntimeException(String.format("location %s does not have any valid version: ", this.location));
+      }
+
+      return res;
+    } catch (IOException ioe) {
+      LOG.error("can not find current version: " + ioe.getMessage(), ioe);
+      throw new RuntimeException(ioe);
+    }
+  }
+
+  @Override
+  public String getCurrentVersion() {
+    return this.currentVersion;
+  }
+  
+  @Override
+  public URI getStoreURI(){
+    return storeURI;
+  }
+  
+  @SuppressWarnings("deprecation")
+  protected Path getVersionRoot(String version){
+    Path versionRoot = new Path(this.location, version);
+    try {
+      if(this.fs.isDirectory(versionRoot)){
```

Comment:
This check will be run almost every time a config is requested. Doing it once should be enough.